sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -912,3 +913,17 @@ case class Deduplicate(
 
   override def output: Seq[Attribute] = child.output
 }
+
+/** A logical plan for setting a barrier of analysis */
+case class AnalysisBarrier(child: LogicalPlan) extends LeafNode {
+  override def output: Seq[Attribute] = child.output
+  override def analyzed: Boolean = true
+  override def isStreaming: Boolean = child.isStreaming
+  override lazy val canonicalized: LogicalPlan = child.canonicalized
+
+  override def find(f: LogicalPlan => Boolean): Option[LogicalPlan] = if (f(this)) {
```

Comment:
if we can't do this, I think we should handle `AnalysisBarrier` in `CacheManager`. Overwriting `find` is too hacky and too far away from `CacheManager`